File: results/2026-04-02_163146/tasks/18-secret-rotation-validator/csharp-script-opus/generated-code/SecretRotationValidator.cs
// Secret Rotation Validator — .NET 10 file-based app
// Run with: dotnet run SecretRotationValidator.cs -- [options]
//
// Validates secrets against their rotation policies, groups by urgency,
// and outputs reports in markdown or JSON format.
//
// Usage:
//   dotnet run SecretRotationValidator.cs -- --config secrets.json [--format markdown|json] [--warning-days 7]
//   dotnet run SecretRotationValidator.cs -- --config secrets.json --format json --warning-days 14

#:package System.Text.Json@9.*

using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.Json;

// --- Entry point ---

var (configPath, format, warningDays) = ParseArgs(args);

try
{
    var secrets = LoadFromFile(configPath);
    var report = Validate(secrets, DateTime.Today, warningDays);

    var output = format switch
    {
        "json" => FormatJson(report, DateTime.Today),
        "markdown" or "md" => FormatMarkdown(report, DateTime.Today),
        _ => throw new ArgumentException($"Unknown format: {format}. Use 'markdown' or 'json'.")
    };

    Console.WriteLine(output);

    // Exit with non-zero if any secrets are expired
    if (report.Expired.Count > 0)
        Environment.Exit(1);
}
catch (ConfigLoadException ex)
{
    Console.Error.WriteLine($"Error loading configuration: {ex.Message}");
    Environment.Exit(2);
}
catch (Exception ex)
{
    Console.Error.WriteLine($"Unexpected error: {ex.Message}");
    Environment.Exit(3);
}

// --- Argument parsing ---

static (string configPath, string format, int warningDays) ParseArgs(string[] args)
{
    string? configPath = null;
    string format = "markdown";
    int warningDays = 7;

    for (int i = 0; i < args.Length; i++)
    {
        switch (args[i])
        {
            case "--config" or "-c":
                if (i + 1 >= args.Length)
                    ExitWithUsage("--config requires a file path argument.");
                configPath = args[++i];
                break;
            case "--format" or "-f":
                if (i + 1 >= args.Length)
                    ExitWithUsage("--format requires a value (markdown or json).");
                format = args[++i].ToLowerInvariant();
                break;
            case "--warning-days" or "-w":
                if (i + 1 >= args.Length || !int.TryParse(args[i + 1], out warningDays) || warningDays < 0)
                    ExitWithUsage("--warning-days requires a non-negative integer.");
                i++;
                break;
            case "--help" or "-h":
                ExitWithUsage(null);
                break;
            default:
                ExitWithUsage($"Unknown argument: {args[i]}");
                break;
        }
    }

    if (configPath is null)
        ExitWithUsage("--config is required.");

    return (configPath!, format, warningDays);
}

static void ExitWithUsage(string? error)
{
    if (error is not null)
        Console.Error.WriteLine($"Error: {error}");

    Console.Error.WriteLine("""
        Usage: dotnet run SecretRotationValidator.cs -- --config <path> [--format markdown|json] [--warning-days <n>]

        Options:
          --config, -c         Path to secrets JSON configuration file (required)
          --format, -f         Output format: markdown (default) or json
          --warning-days, -w   Days before expiry to trigger warning (default: 7)
          --help, -h           Show this help message
        """);
    Environment.Exit(error is null ? 0 : 2);
}

// --- Models ---

record SecretConfig(string Name, DateTime LastRotated, int RotationPolicyDays, List<string> RequiredByServices)
{
    public int DaysUntilExpiry(DateTime asOf) => (int)(LastRotated.AddDays(RotationPolicyDays) - asOf).TotalDays;
    public bool IsExpired(DateTime asOf) => DaysUntilExpiry(asOf) < 0;
    public bool IsInWarningWindow(DateTime asOf, int warningDays)
    {
        var days = DaysUntilExpiry(asOf);
        return days >= 0 && days <= warningDays;
    }
}

enum Urgency { Expired, Warning, Ok }
record ValidationEntry(SecretConfig Secret, Urgency Urgency, int DaysUntilExpiry);
record ValidationReport(List<ValidationEntry> Expired, List<ValidationEntry> Warning, List<ValidationEntry> Ok)
{
    public IEnumerable<ValidationEntry> All => Expired.Concat(Warning).Concat(Ok);
}

// --- Validator ---

static ValidationReport Validate(List<SecretConfig> secrets, DateTime asOf, int warningDays)
{
    var expired = new List<ValidationEntry>();
    var warning = new List<ValidationEntry>();
    var ok = new List<ValidationEntry>();

    foreach (var secret in secrets)
    {
        var daysUntilExpiry = secret.DaysUntilExpiry(asOf);
        Urgency urgency;
        if (secret.IsExpired(asOf)) urgency = Urgency.Expired;
        else if (secret.IsInWarningWindow(asOf, warningDays)) urgency = Urgency.Warning;
        else urgency = Urgency.Ok;

        var entry = new ValidationEntry(secret, urgency, daysUntilExpiry);
        switch (urgency)
        {
            case Urgency.Expired: expired.Add(entry); break;
            case Urgency.Warning: warning.Add(entry); break;
            case Urgency.Ok:      ok.Add(entry); break;
        }
    }
    return new ValidationReport(expired, warning, ok);
}

// --- Config loader ---

class ConfigLoadException : Exception
{
    public ConfigLoadException(string message) : base(message) { }
    public ConfigLoadException(string message, Exception inner) : base(message, inner) { }
}

static List<SecretConfig> LoadFromFile(string filePath)
{
    if (!File.Exists(filePath))
        throw new ConfigLoadException($"Configuration file not found: {filePath}");
    return LoadFromString(File.ReadAllText(filePath));
}

static List<SecretConfig> LoadFromString(string json)
{
    JsonDocument doc;
    try { doc = JsonDocument.Parse(json); }
    catch (JsonException ex) { throw new ConfigLoadException($"Invalid JSON: {ex.Message}", ex); }

    if (!doc.RootElement.TryGetProperty("secrets", out var secretsArray))
        throw new ConfigLoadException("Configuration must contain a 'secrets' array.");

    var results = new List<SecretConfig>();
    for (int i = 0; i < secretsArray.GetArrayLength(); i++)
    {
        var el = secretsArray[i];

        if (!el.TryGetProperty("name", out var nameEl) || nameEl.ValueKind != JsonValueKind.String || string.IsNullOrWhiteSpace(nameEl.GetString()))
            throw new ConfigLoadException($"Secret at index {i}: 'name' is required.");
        var name = nameEl.GetString()!;

        if (!el.TryGetProperty("lastRotated", out var dateEl) || dateEl.ValueKind != JsonValueKind.String)
            throw new ConfigLoadException($"Secret '{name}': 'lastRotated' is required.");
        if (!DateTime.TryParseExact(dateEl.GetString(), "yyyy-MM-dd", CultureInfo.InvariantCulture, DateTimeStyles.None, out var lastRotated))
            throw new ConfigLoadException($"Secret '{name}': 'lastRotated' value '{dateEl.GetString()}' is not a valid date (yyyy-MM-dd).");

        if (!el.TryGetProperty("rotationPolicyDays", out var policyEl) || policyEl.ValueKind != JsonValueKind.Number)
            throw new ConfigLoadException($"Secret '{name}': 'rotationPolicyDays' is required.");
        var policyDays = policyEl.GetInt32();
        if (policyDays <= 0)
            throw new ConfigLoadException($"Secret '{name}': 'rotationPolicyDays' must be positive, got {policyDays}.");

        var services = new List<string>();
        if (el.TryGetProperty("requiredByServices", out var svcEl) && svcEl.ValueKind == JsonValueKind.Array)
            foreach (var s in svcEl.EnumerateArray())
                if (s.ValueKind == JsonValueKind.String) services.Add(s.GetString()!);

        results.Add(new SecretConfig(name, lastRotated, policyDays, services));
    }
    return results;
}

// --- Report formatters ---

static string FormatMarkdown(ValidationReport report, DateTime asOf)
{
    var sb = new StringBuilder();
    var total = report.All.Count();

    sb.AppendLine("# Secret Rotation Report");
    sb.AppendLine($"Generated: {asOf:yyyy-MM-dd}");
    sb.AppendLine();

    if (total == 0)
    {
        sb.AppendLine("No secrets to report.");
        return sb.ToString();
    }

    sb.AppendLine($"**Summary:** {total} secrets — {report.Expired.Count} expired, {report.Warning.Count} warning, {report.Ok.Count} ok");
    sb.AppendLine();

    AppendSection(sb, "Expired", report.Expired);
    AppendSection(sb, "Warning", report.Warning);
    AppendSection(sb, "Ok", report.Ok);

    return sb.ToString();
}

static void AppendSection(StringBuilder sb, string title, List<ValidationEntry> entries)
{
    if (entries.Count == 0) return;
    sb.AppendLine($"## {title}");
    sb.AppendLine();
    sb.AppendLine("| Name | Last Rotated | Policy (days) | Days Until Expiry | Required By |");
    sb.AppendLine("|------|-------------|---------------|-------------------|-------------|");
    foreach (var e in entries)
    {
        var svcs = string.Join(", ", e.Secret.RequiredByServices);
        sb.AppendLine($"| {e.Secret.Name} | {e.Secret.LastRotated:yyyy-MM-dd} | {e.Secret.RotationPolicyDays} | {e.DaysUntilExpiry} | {svcs} |");
    }
    sb.AppendLine();
}

static string FormatJson(ValidationReport report, DateTime asOf)
{
    var data = new
    {
        reportDate = asOf.ToString("yyyy-MM-dd"),
        summary = new
        {
            total = report.All.Count(),
            expired = report.Expired.Count,
            warning = report.Warning.Count,
            ok = report.Ok.Count
        },
        expired = report.Expired.Select(ToJsonEntry),
        warning = report.Warning.Select(ToJsonEntry),
        ok = report.Ok.Select(ToJsonEntry)
    };
    return JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true, PropertyNamingPolicy = JsonNamingPolicy.CamelCase });
}

static object ToJsonEntry(ValidationEntry entry) => new
{
    name = entry.Secret.Name,
    lastRotated = entry.Secret.LastRotated.ToString("yyyy-MM-dd"),
    rotationPolicyDays = entry.Secret.RotationPolicyDays,
    daysUntilExpiry = entry.DaysUntilExpiry,
    urgency = entry.Urgency.ToString().ToLowerInvariant(),
    requiredByServices = entry.Secret.RequiredByServices
};

File: results/2026-04-08_192624/tasks/18-secret-rotation-validator/default-opus/generated-code/secret_rotation_validator.py
#!/usr/bin/env python3
"""
Secret Rotation Validator

Reads a JSON configuration of secrets with metadata, identifies secrets that are
expired or expiring within a configurable warning window, and generates a rotation
report grouped by urgency (expired, warning, ok).

Supports output formats: JSON, markdown table.

Usage:
    python3 secret_rotation_validator.py --config secrets.json [--warning-days 14] [--format json|markdown] [--reference-date YYYY-MM-DD]
"""

import argparse
import json
import sys
from datetime import datetime, timedelta


def parse_args(argv=None):
    """Parse command-line arguments."""
    parser = argparse.ArgumentParser(description="Validate secret rotation status")
    parser.add_argument(
        "--config", required=True, help="Path to JSON config file with secret metadata"
    )
    parser.add_argument(
        "--warning-days",
        type=int,
        default=14,
        help="Number of days before expiry to trigger a warning (default: 14)",
    )
    parser.add_argument(
        "--format",
        choices=["json", "markdown"],
        default="json",
        help="Output format: json or markdown (default: json)",
    )
    parser.add_argument(
        "--reference-date",
        default=None,
        help="Reference date for calculations (YYYY-MM-DD). Defaults to today.",
    )
    return parser.parse_args(argv)


def load_config(config_path):
    """Load and validate the secrets configuration file."""
    try:
        with open(config_path, "r") as f:
            data = json.load(f)
    except FileNotFoundError:
        print(f"Error: Config file not found: {config_path}", file=sys.stderr)
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"Error: Invalid JSON in config file: {e}", file=sys.stderr)
        sys.exit(1)

    if "secrets" not in data or not isinstance(data["secrets"], list):
        print("Error: Config must contain a 'secrets' array", file=sys.stderr)
        sys.exit(1)

    required_fields = ["name", "last_rotated", "rotation_policy_days", "required_by"]
    for i, secret in enumerate(data["secrets"]):
        for field in required_fields:
            if field not in secret:
                print(
                    f"Error: Secret at index {i} missing required field '{field}'",
                    file=sys.stderr,
                )
                sys.exit(1)

    return data["secrets"]


def classify_secret(secret, reference_date, warning_days):
    """
    Classify a single secret into expired, warning, or ok.

    Returns a dict with the secret info plus:
      - status: 'expired', 'warning', or 'ok'
      - expires_on: the computed expiration date string
      - days_until_expiry: integer (negative means already expired)
    """
    last_rotated = datetime.strptime(secret["last_rotated"], "%Y-%m-%d").date()
    policy_days = secret["rotation_policy_days"]
    expires_on = last_rotated + timedelta(days=policy_days)
    days_until_expiry = (expires_on - reference_date).days

    if days_until_expiry < 0:
        status = "expired"
    elif days_until_expiry <= warning_days:
        status = "warning"
    else:
        status = "ok"

    return {
        "name": secret["name"],
        "last_rotated": secret["last_rotated"],
        "rotation_policy_days": policy_days,
        "required_by": secret["required_by"],
        "expires_on": expires_on.isoformat(),
        "days_until_expiry": days_until_expiry,
        "status": status,
    }


def generate_report(secrets, reference_date, warning_days):
    """
    Generate a full rotation report from a list of secrets.

    Returns a dict with:
      - reference_date: the date used for calculations
      - warning_days: the warning window
      - summary: counts per status
      - groups: { expired: [...], warning: [...], ok: [...] }
    """
    classified = [classify_secret(s, reference_date, warning_days) for s in secrets]

    groups = {"expired": [], "warning": [], "ok": []}
    for item in classified:
        groups[item["status"]].append(item)

    # Sort expired by most overdue first, warning by soonest first
    groups["expired"].sort(key=lambda x: x["days_until_expiry"])
    groups["warning"].sort(key=lambda x: x["days_until_expiry"])
    groups["ok"].sort(key=lambda x: x["days_until_expiry"])

    return {
        "reference_date": reference_date.isoformat(),
        "warning_days": warning_days,
        "summary": {
            "total": len(classified),
            "expired": len(groups["expired"]),
            "warning": len(groups["warning"]),
            "ok": len(groups["ok"]),
        },
        "groups": groups,
    }


def format_json(report):
    """Format report as JSON string."""
    return json.dumps(report, indent=2)


def format_markdown(report):
    """Format report as a markdown table grouped by urgency."""
    lines = []
    lines.append(f"# Secret Rotation Report")
    lines.append(f"")
    lines.append(f"**Reference Date:** {report['reference_date']}")
    lines.append(f"**Warning Window:** {report['warning_days']} days")
    lines.append(f"")
    lines.append(f"## Summary")
    lines.append(f"")
    s = report["summary"]
    lines.append(f"| Status | Count |")
    lines.append(f"|--------|-------|")
    lines.append(f"| Expired | {s['expired']} |")
    lines.append(f"| Warning | {s['warning']} |")
    lines.append(f"| OK | {s['ok']} |")
    lines.append(f"| **Total** | **{s['total']}** |")
    lines.append(f"")

    for group_name, label in [
        ("expired", "EXPIRED"),
        ("warning", "WARNING"),
        ("ok", "OK"),
    ]:
        items = report["groups"][group_name]
        lines.append(f"## {label} ({len(items)})")
        lines.append(f"")
        if items:
            lines.append(
                f"| Name | Last Rotated | Policy (days) | Expires On | Days Until Expiry | Required By |"
            )
            lines.append(
                f"|------|-------------|---------------|------------|-------------------|-------------|"
            )
            for item in items:
                services = ", ".join(item["required_by"])
                lines.append(
                    f"| {item['name']} | {item['last_rotated']} | {item['rotation_policy_days']} | {item['expires_on']} | {item['days_until_expiry']} | {services} |"
                )
        else:
            lines.append("No secrets in this category.")
        lines.append(f"")

    return "\n".join(lines)


def main(argv=None):
    """Main entry point."""
    args = parse_args(argv)

    if args.reference_date:
        try:
            reference_date = datetime.strptime(args.reference_date, "%Y-%m-%d").date()
        except ValueError:
            print(
                f"Error: Invalid reference date format: {args.reference_date}. Use YYYY-MM-DD.",
                file=sys.stderr,
            )
            sys.exit(1)
    else:
        reference_date = datetime.now().date()

    secrets = load_config(args.config)
    report = generate_report(secrets, reference_date, args.warning_days)

    if args.format == "json":
        output = format_json(report)
    else:
        output = format_markdown(report)

    print(output)

    # Exit with non-zero if any secrets are expired
    if report["summary"]["expired"] > 0:
        sys.exit(2)


if __name__ == "__main__":
    main()
